### PR TITLE
feat: implement DP-101..DP-104 percolation policy core

### LINKED ISSUE #230: Define surfaced-item schema and policy input contract
> Source catalog: docs/roadmap/v0.8.0-dynamic-percolation-issues.md

# Define surfaced-item schema and policy input contract

1. Title
- Define surfaced-item schema and policy input contract

2. Parent epic
- #225

3. Problem
- ranking inputs are implicit across modules, making behavior hard to reason about.

4. Why this matters
- deterministic policy and explainability require a stable typed contract.

5. Proposal
- add `src/percolation/types.ts` with `NormalizedSignal`, `SurfacedItem`, and `PercolationDecision`.

6. Implementation notes
- likely touched: `src/extension.ts`, `src/types.ts`, new `src/percolation/*`
- keep schema strictly local and deterministic
- add migration-safe defaults for missing fields

7. Acceptance criteria
- policy input/output interfaces documented and exported
- no behavior change yet; adapter compiles with existing summary flow

8. Test plan
- unit tests for schema normalization and defaulting behavior

9. Telemetry / metrics impact
- none required in this issue

10. Risks / edge cases
- schema churn if downstream requirements not captured early

11. Dependencies
- none

12. Suggested labels
- `area:policy`, `v0.8.0`

13. Suggested milestone
- `v0.8.0-dynamic-percolation-core`

14. Mockup reference
- `docs/ux/dynamic-percolation-mockups.md#signal---policy---surface-flow`
## Linked Epic
- #225

### LINKED ISSUE #231: Implement deterministic percolation ranking pipeline v1
> Source catalog: docs/roadmap/v0.8.0-dynamic-percolation-issues.md

# Implement deterministic percolation ranking pipeline v1

1. Title
- Implement deterministic percolation ranking pipeline v1

2. Parent epic
- #225

3. Problem
- TaCoS lacks one ranking function for all candidate surfaced items.

4. Why this matters
- ranking consistency is required for calm, predictable UX.

5. Proposal
- create `rankCandidates()` with fixed weights for urgency, actionability, continuity, novelty, and interrupt cost.

6. Implementation notes
- likely touched: new `src/percolation/ranking.ts`, `src/extension.ts`
- output stable sort order with tie-break keys
- include score breakdown for explainability

7. Acceptance criteria
- same input state returns identical ranking order
- ranking result consumed by panel/status/notification selector

8. Test plan
- deterministic ranking unit tests with snapshot fixtures

9. Telemetry / metrics impact
- add optional score-band counters behind feature flag

10. Risks / edge cases
- overweighting novelty can degrade continuity

11. Dependencies
- DP-101

12. Suggested labels
- `area:policy`, `v0.8.0`, `performance`

13. Suggested milestone
- `v0.8.0-dynamic-percolation-core`

14. Mockup reference
- `docs/ux/dynamic-percolation-mockups.md#signal---policy---surface-flow`
## Linked Epic
- #225

### LINKED ISSUE #232: Add suppression gates for quiet, cooldown, no-change, and noise budget
> Source catalog: docs/roadmap/v0.8.0-dynamic-percolation-issues.md

# Add suppression gates for quiet, cooldown, no-change, and noise budget

1. Title
- Add suppression gates for quiet, cooldown, no-change, and noise budget

2. Parent epic
- #225

3. Problem
- suppression exists across multiple code paths and is not uniformly applied to all surfaced candidates.

4. Why this matters
- inconsistent suppression increases interruption risk.

5. Proposal
- centralize suppression evaluation in policy layer and return canonical suppression reason.

6. Implementation notes
- likely touched: `src/noiseControl.ts`, `src/extension.ts`, new `src/percolation/suppression.ts`
- preserve existing quiet/snooze semantics
- explicitly model `no-change` suppression

7. Acceptance criteria
- suppression reasons are deterministic and serializable
- nudges, notifications, and panel emphasis use the same gate results

8. Test plan
- unit tests for suppression precedence and next-eligible timestamp

9. Telemetry / metrics impact
- track suppression counts by reason

10. Risks / edge cases
- over-suppression can hide urgent blockers

11. Dependencies
- DP-101

12. Suggested labels
- `area:policy`, `v0.8.0`

13. Suggested milestone
- `v0.8.0-dynamic-percolation-core`

14. Mockup reference
- `docs/ux/dynamic-percolation-mockups.md#notification-vs-statusbar-decision-flow`
## Linked Epic
- #225

### LINKED ISSUE #233: Add unified explainability payload (why surfaced / why suppressed)
> Source catalog: docs/roadmap/v0.8.0-dynamic-percolation-issues.md

# Add unified explainability payload (why surfaced / why suppressed)

1. Title
- Add unified explainability payload (why surfaced / why suppressed)

2. Parent epic
- #225

3. Problem
- current explainability is split across trust cue and nudge-specific text.

4. Why this matters
- users need one coherent answer for "why now".

5. Proposal
- attach explanation payload to policy decision and render in panel drill-down.

6. Implementation notes
- likely touched: `src/extension.ts`, `src/webview/panelFragments.ts`, `src/webview/panelCards.ts`
- include evidence IDs, suppression reason, confidence, and missing-signal lines

7. Acceptance criteria
- every surfaced primary item has explanation data
- suppressed decisions expose reason text

8. Test plan
- unit tests for explanation payload formatting
- integration test for "Why am I seeing this?" visibility

9. Telemetry / metrics impact
- add `whySurfacedOpenRate` tracking hook (actual metric plumbing in E5)

10. Risks / edge cases
- explanation verbosity can create clutter

11. Dependencies
- DP-102, DP-103

12. Suggested labels
- `area:policy`, `area:trust-privacy`, `v0.8.0`

13. Suggested milestone
- `v0.8.0-dynamic-percolation-core`

14. Mockup reference
- `docs/ux/dynamic-percolation-mockups.md#trust--privacy--evidence-quick-drill-down`
## Linked Epic
- #225

### PR DESCRIPTION
**Pull request by jkordish**
## Summary
This PR implements the first percolation policy core slices under epic #225:
- DP-101 (`#230`): typed percolation contracts + deterministic normalization adapters
- DP-102 (`#231`): deterministic ranking pipeline with fixed weights and score breakdowns
- DP-103 (`#232`): shared suppression gates (quiet/cooldown/no-change/noise-budget) reused across surfaces
- DP-104 (`#233`): unified explainability payload rendered in Trust Center "Why am I seeing this?" drill-down

## What Changed
- Added new percolation modules:
  - `src/percolation/types.ts`
  - `src/percolation/ranking.ts`
  - `src/percolation/suppression.ts`
  - `src/percolation/explainability.ts`
- Integrated ranking/suppression/explainability into surface selection and panel rendering in `src/extension.ts`
- Added explainability open hook (`whySurfacedOpened`) and local metrics counters/exports for suppression reasons and explainability opens
- Updated companion/webview plumbing and tests accordingly

## Verification
- `npm run compile`
- `npm run lint`
- `npm test -- percolationTypes.test.ts percolationRanking.test.ts percolationSuppression.test.ts percolationExplainability.test.ts companionNudges.test.ts metrics.test.ts panelCards.test.ts webviewMessages.test.ts`

## Issue Links
- Parent epic: #225
- Closes #230
- Closes #231
- Closes #232
- Closes #233

### CODE REVIEW

File: src/percolation/types.ts [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+  return `signal:${fallbackKind}`;
+}
+
+function resolveItemId(item: Partial<SurfacedItem>, fallbackKind: SurfacedItemKind): string {
+  if (item.id && item.id.trim().length > 0) {
+    return item.id.trim();
+  }
+
+  const titleToken = item.title?.trim().toLowerCase().replace(/[^a-z0-9]+/gu, '-') ?? 'item';
```
[the remark this answers is not in the document]
Comment (jkordish) [reply src/percolation/types.ts]: Addressed in commit 9461bbc. `resolveItemId()` now slugifies from normalized title text, trims edge dashes, and falls back to `item` when the slug is empty, preventing IDs like `item:status:` and reducing collision risk.

**Comment by jkordish** [comment]
@copilot review

File: src/percolation/types.ts [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    secondary,
+    surfaced,
+    suppressionReason: decision.suppressionReason,
+    nextEligibleAt:
+      typeof decision.nextEligibleAt === 'number' && Number.isFinite(decision.nextEligibleAt)
+        ? Math.max(0, Math.floor(decision.nextEligibleAt))
+        : undefined,
+    explain: {
+      summary: decision.explain?.summary?.trim() ?? '',
+      reasons,
```
[the remark this answers is not in the document]
Comment (jkordish) [reply src/percolation/types.ts]: Addressed in follow-up PR #261 (commit 3efe7ad), which hardens percolation normalizers against non-string runtime payload values and adds regression tests.
